Configuration view improvements

ktkization opened the pull request.

ktkization committed:
Remove default suggested replies for new tags

ktkization committed:
Remove tag from dropdown after adding; Add close button for add tag modal

ktkization committed:
Mark newly added tag as active

ktkization committed:
Drag and drop adding for new response text

marianamarasoiu commented on webapp/lib/controller.dart:
```diff
@@ -284,33 +233,52 @@ void command(UIAction action, Data actionData) {
-  view.contentView.configurationView.tagList.renderTagList(tagData.keys.toList());
-  view.contentView.configurationView.tagResponses.renderResponses(tagData.keys.toList().first, tagData.values.toList().first);
+void showConfigurationView([String selectedTag, Map<String, Map<String, List<String>>> filteredTagData]) {
+  Map<String, bool> tags = new Map.fromIterable(tagData.keys.toList(),
```
should it be `filteredTagData` instead of `tagData` here?

marianamarasoiu replied:
Oh, I think I see what you're doing.

In general, and this function is a good example of that, it would be good to make functions self-sufficient - ie. to have access only to the data that gets passed to it as arguments, rather than drawing on other data that is not a constant variable (such as `tagData`). Could the code that handles `UIAction.goToConfigurator` (line 220) prepare `filteredTagData` and then pass it on to this method?

ktkization replied:
I've prepared the filtered data when we `UIAction.goToConfigurator` is executed.

marianamarasoiu replied:
That's good progress, I think we can work this method even further - basically pass it only the data that it needs, which are: the selected tag, the tag list, and the replies for that tag. I suggest you change the method to take these three things as the arguments, and then adjusted the calling code to prepare this data for it (and the code inside it etc.).

ktkization replied:
Refactored it to take those 2 params

marianamarasoiu commented on webapp/lib/controller.dart:
```diff
@@ -284,33 +233,52 @@ void command(UIAction action, Data actionData) {
-  selectedConfigurationTag = selectedTag;
-  var filteredTagResponses = Map<String, Map<String, List<String>>>.from(tagData)..removeWhere((k, v) => !k.contains(selectedTag));
-  view.contentView.configurationView.tagResponses.renderResponses(filteredTagResponses.keys.toList().first, filteredTagResponses.values.toList().first);
+  var filteredTagData = Map<String, Map<String, List<String>>>.from(tagData)..removeWhere((k, v) => !k.contains(selectedTag));
```
`filtered` is not a very good name here - what you're doing is taking the messages corresponding to the selected tag. how about `messagesForTag`?
Also, wouldn't it be more straightforward to get this data by doing:
```suggestion
  var filteredTagData = tagData[selectedTag];
```
(which you also don't really need to wrap in a method)

ktkization replied:
Done some data refactoring as per your previous comment. I've used `selectedTagResponses`

marianamarasoiu commented on webapp/lib/controller.dart:
```diff
@@ -284,33 +233,52 @@ void command(UIAction action, Data actionData) {
 void addNewConfigurationTag(String tagToAdd) {
-  tagData[tagToAdd] = addtitionalTagData[tagToAdd];
-  showConfigurationView();
+  var availableLanguages =  tagData.values.toList().map((d) => d.keys.toList()).expand((d) => d).toSet();
```
we should have a separate list of languages, rather than reading them from the `tagData`

ktkization replied:
I've prepared a separate language list

marianamarasoiu commented on webapp/lib/view.dart:
```diff
@@ -211,26 +212,40 @@ class ConfigurationViewTagListPartial {
+      tagItem.onDrop.listen((event) {
+        event.preventDefault();
+        var dropTarget = (event.target as Element);
+        dropTarget.style.backgroundColor = '#ffffff';
```
we shouldn't try to directly set css values here - we should use css classes instead

marianamarasoiu replied:
this applies to all the other places in this file, esp. around `backgroundColor`

ktkization replied:
Adding the CSS classes doesn't show the desired CSS effect instantly. I don't know why. That's why I'm applying the CSS directly.

marianamarasoiu reviewed (commented):
thanks @ktkization 
A first review pass with a couple of coding suggestions, but I think there's a functionality change that's needed before further review.

The user should only be able to drag a full row of messages (so message+translation(s)), but not individual messages (otherwise we need to re-translate messages, or we get messages under the wrong language). I think this could work by having a little icon on the left side of each row, and use that as a "drag handle". I also think we don't want to immediately load the view of the new tag, but instead maintain the tag we're currently looking at.

ktkization committed:
Add drag and drop icon for response rows

ktkization committed:
Drag and drop for entire response row

ktkization committed:
Refactor how data is passed around

ktkization committed:
Fix syntax error

ktkization commented:
@marianamarasoiu I've refactored the code to be able to drag and drop a full row instead and also refactored how data is passed around as per your first comment

ktkization committed:
Replace incline styling with classes

marianamarasoiu commented on webapp/lib/controller.dart:
```diff
@@ -4,7 +4,102 @@ import 'logger.dart';
+  }
+}
+
+Set<String> get configurationReponseLanguageData { return {'English', 'Somali'};}
```
I agree with the rationale of having the languages be a set b/c it gives us assurances that a language doesn't appear twice, but unfortunately it's not going to work here because the order is going to be important and user defined. So I suggest we use a list instead, and we can assume that the languages will be different (as a note for the future: perhaps when we add the project configuration page we can actually check user input to make sure they are different)

ktkization replied:
I see how that can be a problem. Changed it to a List

marianamarasoiu commented on webapp/lib/controller.dart:
```diff
@@ -4,7 +4,102 @@ import 'logger.dart';
+  }
+}
+
+Set<String> get configurationReponseLanguageData { return {'English', 'Somali'};}
```
you can also rewrite this to use the `=>` which is the dart recommended way of defining single line functions:
```suggestion
List<String> get configurationReponseLanguageData => ['English', 'Somali'];
```

marianamarasoiu commented on webapp/lib/controller.dart:
```diff
@@ -4,7 +4,102 @@ import 'logger.dart';
+
+Set<String> get configurationReponseLanguageData { return {'English', 'Somali'};}
+
+Set<String> get configurationTags { return {'addtional Tag 1', 'addtional Tag 2', 'addtional Tag 3', 'addtional Tag 4', 'addtional Tag 5'};}
```
Same suggestion as above to use `=>` syntax
```suggestion
Set<String> get configurationTags => {'addtional Tag 1', 'addtional Tag 2', 'addtional Tag 3', 'addtional Tag 4', 'addtional Tag 5'};
```

marianamarasoiu commented on webapp/lib/controller.dart:
```diff
@@ -4,7 +4,102 @@ import 'logger.dart';
+
+Set<String> get configurationTags { return {'addtional Tag 1', 'addtional Tag 2', 'addtional Tag 3', 'addtional Tag 4', 'addtional Tag 5'};}
+
+Map<String, Map<String, List<String>>> get configurationData {
+  return {
```
```suggestion
Map<String, Map<String, List<String>>> get configurationData =>
  {
```

marianamarasoiu commented on webapp/lib/controller.dart:
```diff
@@ -150,167 +245,46 @@ Map<String, Map<String, List<String>>> tagData = {
-  view.contentView.configurationView.tagList.renderTagList(tagData.keys.toList());
-  view.contentView.configurationView.tagResponses.renderResponses(tagData.keys.toList().first, tagData.values.toList().first);
+void showConfigurationView(String selectedTag, Map<String, Map<String, List<String>>> tagData) {
+  if (selectedTag == null) selectedTag = tagData.keys.toList().first;
```
in general in Dart, passing thing around that are `null` is a bit frowned upon. Null is often an indication that there is a problem (E.g that something hasn't been initialised properly).

I see that the only place where selectedTag is passed in as null is when handling the `UIAction.goToConfigurator`. I would recommend that you compute the selected tag there the same way (ie. as the first one in the list), and then pass it to this method, instead of having this if here.

ktkization replied:
Understood. Computed the `selectedTag` beforehand

marianamarasoiu commented on webapp/lib/view.dart:
```diff
@@ -211,26 +212,42 @@ class ConfigurationViewTagListPartial {
+      tagItem.onDragEnter.listen((event) => event.preventDefault());
+      tagItem.onDragOver.listen((event) {
+        event.preventDefault();
+        (event.target as Element).classes.add('configure-package__tag-item-active');
```
the `onDragOver` event gets triggered every few hundred milliseconds, resulting in this class being added lots of times (or at least attempted to be added lots of times). Shouldn't this class change be done `onDragEnter` instead?

ktkization replied:
Ok. Moved the logic to `onDragEnter`

marianamarasoiu commented on webapp/lib/view.dart:
```diff
@@ -211,26 +212,42 @@ class ConfigurationViewTagListPartial {
+        event.preventDefault();
+        (event.target as Element).classes.add('configure-package__tag-item-active');
+      });
+      tagItem.onDragLeave.listen((event) => (event.target as Element).classes.remove('configure-package__tag-item-active'));
```
it might be good to separate the style between selection, ie. `tag-item-active` (which should probably be renamed to `tag-item--active` since active is a "modifier") and acting as a drop target. (the difference can be just conceptual, even though the background colour for now can be the same). So I suggest that you use a different class here, for example `tag-item--drop-target`

ktkization replied:
Can I change this in the PR I was supposed to do the other BEM fixes?

marianamarasoiu replied:
sure, yes, that makes sense

marianamarasoiu commented on webapp/lib/view.dart:
```diff
@@ -211,26 +212,42 @@ class ConfigurationViewTagListPartial {
+        dropTarget.classes.remove('configure-package__tag-item-active');
+        if (dropTarget.classes.contains('configure-package__tag-item')) {
+          var responseData = jsonDecode(event.dataTransfer.getData("Text"));
+          responseData.forEach((language, text) {
+            controller.command(controller.UIAction.addConfigurationResponseEntries,
+              new controller.ConfigurationResponseData(parentTag: dropTarget.text, language: language, text: text));
```
As discussed earlier - this should be done in a row at once, rather than one text message at a time - nothing to do now about it, but to be picked up in the future PR that adjusts the model

ktkization replied:
ok

marianamarasoiu commented on webapp/web/configuration.css:
```diff
@@ -119,6 +119,31 @@
+    position: relative;
+}
+
+.configure-package__tag-responses-item-drag {
```
nit: rename this to
```suggestion
.configure-package__tag-responses-item-drag-handle {
```

marianamarasoiu commented on webapp/web/configuration.css:
```diff
@@ -119,6 +119,31 @@
+    right: 15px;
+}
+
+.configure-package__tag-responses-item-drag-0 {
```
nit:
```suggestion
.configure-package__tag-responses-item-drag-handle-0 {
```

marianamarasoiu reviewed (approved):
thanks @ktkization - here's a bunch of suggestions, but most of them fairly minor. I think ok to merge after you address these, and we'll pick up the more substantial changes we talked about earlier for future PR rather than trying to get them all fixed now.

ktkization committed:
Code improvement

Co-authored-by: Mariana Mărășoiu <mariana.marasoiu@gmail.com>

ktkization committed:
Code improvement

Co-authored-by: Mariana Mărășoiu <mariana.marasoiu@gmail.com>

ktkization committed:
Rename CSS class

Co-authored-by: Mariana Mărășoiu <mariana.marasoiu@gmail.com>

ktkization committed:
Rename CSS class

Co-authored-by: Mariana Mărășoiu <mariana.marasoiu@gmail.com>

ktkization committed:
Code improvement

Co-authored-by: Mariana Mărășoiu <mariana.marasoiu@gmail.com>

ktkization committed:
Rename CSS class for drag and drop

ktkization committed:
Remove extra bracket

ktkization committed:
Change onDragOver for tagItem to onDragEnter and rename CSS class

ktkization committed:
Change configurationResponseLanguages to from Set to List

ktkization committed:
Change parameter list for showConfigurationView()

marianamarasoiu reviewed (approved):
thanks Kenneth!
